cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/DirectRuntimeRequestValidator.java

```diff
@@ -73,51 +75,66 @@
+          return Optional.ofNullable(getRunRecordStatusForProgramsInNonEndState(programRunId));
         }
       });
   }
 
   @Override
-  public void validate(ProgramRunId programRunId, HttpRequest request) throws BadRequestException {
+  public RunRecordStatus checkProgramRunStatus(ProgramRunId programRunId, HttpRequest request) throws
+    BadRequestException {
     accessEnforcer.enforce(programRunId, authenticationContext.getPrincipal(), StandardPermission.GET);
-    boolean exists;
+    RunRecordStatus runRecordStatus;
     try {
-      exists = programRunsCache.get(programRunId);
+      runRecordStatus = programRunsCache.get(programRunId)
+        .orElseThrow(() -> new BadRequestException("Program run " + programRunId + " is not valid"));
+    } catch (BadRequestException e) {
+      throw e;
     } catch (Exception e) {
       throw new ServiceUnavailableException(Constants.Service.RUNTIME, e);
     }
-    if (!exists) {
-      throw new BadRequestException("Program run " + programRunId + " is not valid");
-    }
+    return runRecordStatus;
   }
 
-  /**
-   * Checks if the given {@link ProgramRunId} is valid.
-   */
-  private boolean isValid(ProgramRunId programRunId) throws IOException, UnauthorizedException {
-    RunRecordDetail runRecord = TransactionRunners.run(txRunner, context -> {
-      return AppMetadataStore.create(context).getRun(programRunId);
-    }, IOException.class);
-
-    if (runRecord != null) {
-      return !runRecord.getStatus().isEndState();
-    }
-    // If it is not found in the local store, which should be very rare, try to fetch the run record remotely.
+  @Nullable
+  private RunRecordStatus getRunRecordStatusForProgramsInNonEndState(ProgramRunId programRunId)
+    throws IOException, UnauthorizedException {
     try {
+      RunRecordDetail runRecord = TransactionRunners.run(txRunner, context -> {
+        return AppMetadataStore.create(context).getRun(programRunId);
+      }, IOException.class);
+
+      if (runRecord != null) {
+        return runRecord.getStatus().isEndState() ? null : getValidRunRecordStatus(runRecord);
+      }
+      // If it is not found in the local store, which should be very rare, try to fetch the run record remotely.
       LOG.info("Remotely fetching program run details for {}", programRunId);
       runRecord = runRecordFetcher.getRunRecordMeta(programRunId);
       // Try to update the local store
       insertRunRecord(programRunId, runRecord);
-      return !runRecord.getStatus().isEndState();
+      return runRecord.getStatus().isEndState() ? null : getValidRunRecordStatus(runRecord);
     } catch (NotFoundException e) {
-      return false;
+      return null;
     }
   }
 
+  private RunRecordStatus getValidRunRecordStatus(RunRecordDetail runRecord) {
+    ProgramRunStatus programRunStatus = runRecord.getStatus();
+    RunRecordStatus.Builder runRecordStatusBuilder = RunRecordStatus.builder().setProgramRunStatus(programRunStatus);
```

Comment:
If both parameters are set at the same time, do you think builder pattern is needed in this case?